File: integrations/sie_langchain/src/sie_langchain/extractors.py
"""SIE extraction tool for LangChain.

Provides entity, relation, classification, and object extraction using SIE's extract endpoint.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from langchain_core.tools import BaseTool
from pydantic import ConfigDict, Field
from sie_sdk.types import Item

if TYPE_CHECKING:
    from langchain_core.callbacks import CallbackManagerForToolRun
    from sie_sdk import SIEAsyncClient, SIEClient


class SIEExtractor(BaseTool):
    """LangChain tool for extraction using SIE.

    Wraps SIEClient.extract() to implement BaseTool for use in agents.
    Returns entities, relations, classifications, and detected objects.

    Example:
        >>> extractor = SIEExtractor(
        ...     base_url="http://localhost:8080",
        ...     model="urchade/gliner_multi-v2.1",
        ...     labels=["person", "organization", "location"],
        ... )
        >>> result = extractor.invoke("John Smith works at Acme Corp in NYC")
        >>> result["entities"]  # list of entity dicts
        >>> result["relations"]  # list of relation dicts

    Args:
        base_url: URL of the SIE server.
        model: Extraction model name/ID.
        labels: Labels to extract (entity types, relation types, or classification labels).
        client: Optional pre-configured SIEClient instance.
        async_client: Optional pre-configured SIEAsyncClient instance.
        options: Runtime options dict for model adapter overrides.
        gpu: Target GPU type for routing (e.g., "l4", "a100-80gb").
        timeout_s: Request timeout in seconds.
    """

    # Pydantic v2 config
    model_config = ConfigDict(arbitrary_types_allowed=True, extra="forbid")

    name: str = "sie_extract"
    description: str = (
        "Extract structured information from text. "
        "Input should be text to analyze. "
        "Returns entities, relations, classifications, and detected objects."
    )

    base_url: str = Field(default="http://localhost:8080")
    model: str = Field(default="urchade/gliner_multi-v2.1")
    labels: list[str] = Field(default_factory=lambda: ["person", "organization", "location"])
    options: dict[str, Any] | None = Field(default=None)
    gpu: str | None = Field(default=None)
    timeout_s: float = Field(default=180.0)

    # Private attributes for clients
    _client: Any = None
    _async_client: Any = None

    def __init__(
        self,
        *,
        base_url: str = "http://localhost:8080",
        model: str = "urchade/gliner_multi-v2.1",
        labels: list[str] | None = None,
        client: SIEClient | None = None,
        async_client: SIEAsyncClient | None = None,
        options: dict[str, Any] | None = None,
        gpu: str | None = None,
        timeout_s: float = 180.0,
        **kwargs: Any,
    ) -> None:
        """Initialize SIE extractor."""
        if labels is None:
            labels = ["person", "organization", "location"]
        super().__init__(
            base_url=base_url, model=model, labels=labels, options=options, gpu=gpu, timeout_s=timeout_s, **kwargs
        )
        self._client = client
        self._async_client = async_client

    @property
    def client(self) -> SIEClient:
        """Get or create the sync SIEClient."""
        if self._client is None:
            from sie_sdk import SIEClient

            self._client = SIEClient(
                self.base_url,
                timeout_s=self.timeout_s,
                gpu=self.gpu,
                options=self.options,
            )
        return self._client

    @property
    def async_client(self) -> SIEAsyncClient:
        """Get or create the async SIEClient."""
        if self._async_client is None:
            from sie_sdk import SIEAsyncClient

            self._async_client = SIEAsyncClient(
                self.base_url,
                timeout_s=self.timeout_s,
                gpu=self.gpu,
                options=self.options,
            )
        return self._async_client

    def _run(
        self,
        text: str,
        run_manager: CallbackManagerForToolRun | None = None,  # noqa: ARG002
    ) -> dict[str, list[dict[str, Any]]]:
        """Extract structured information from text.

        Args:
            text: Text to extract from.
            run_manager: Optional callback manager (not used).

        Returns:
            Dict with entities, relations, classifications, and objects lists.
        """
        result = self.client.extract(
            self.model,
            Item(text=text),
            labels=self.labels,
        )

        return self._format_result(result)

    async def _arun(
        self,
        text: str,
        run_manager: CallbackManagerForToolRun | None = None,  # noqa: ARG002
    ) -> dict[str, list[dict[str, Any]]]:
        """Async extract structured information from text.

        Args:
            text: Text to extract from.
            run_manager: Optional callback manager (not used).

        Returns:
            Dict with entities, relations, classifications, and objects lists.
        """
        result = await self.async_client.extract(
            self.model,
            Item(text=text),
            labels=self.labels,
        )

        return self._format_result(result)

    def _format_result(self, result: object) -> dict[str, list[dict[str, Any]]]:
        """Format extraction result into multi-type dict.

        Args:
            result: ExtractResult from SIE.

        Returns:
            Dict with entities, relations, classifications, and objects.
        """

        def _get(key: str) -> list:
            if isinstance(result, dict):
                return result.get(key, [])
            return getattr(result, key, [])

        entities = []
        for e in _get("entities"):
            if isinstance(e, dict):
                entities.append(
                    {
                        "text": e.get("text", ""),
                        "label": e.get("label", ""),
                        "score": float(e.get("score", 0.0)),
                        "start": int(e.get("start", 0)),
                        "end": int(e.get("end", 0)),
                    }
                )
            else:
                entities.append(
                    {
                        "text": getattr(e, "text", ""),
                        "label": getattr(e, "label", ""),
                        "score": float(getattr(e, "score", 0.0)),
                        "start": int(getattr(e, "start", 0)),
                        "end": int(getattr(e, "end", 0)),
                    }
                )

        relations = []
        for r in _get("relations"):
            if isinstance(r, dict):
                relations.append(
                    {
                        "head": r.get("head", ""),
                        "tail": r.get("tail", ""),
                        "relation": r.get("relation", ""),
                        "score": float(r.get("score", 0.0)),
                    }
                )
            else:
                relations.append(
                    {
                        "head": getattr(r, "head", ""),
                        "tail": getattr(r, "tail", ""),
                        "relation": getattr(r, "relation", ""),
                        "score": float(getattr(r, "score", 0.0)),
                    }
                )

        classifications = []
        for c in _get("classifications"):
            if isinstance(c, dict):
                classifications.append({"label": c.get("label", ""), "score": float(c.get("score", 0.0))})
            else:
                classifications.append({"label": getattr(c, "label", ""), "score": float(getattr(c, "score", 0.0))})

        objects = []
        for o in _get("objects"):
            if isinstance(o, dict):
                objects.append(
                    {"label": o.get("label", ""), "score": float(o.get("score", 0.0)), "bbox": o.get("bbox", [])}
                )
            else:
                objects.append(
                    {
                        "label": getattr(o, "label", ""),
                        "score": float(getattr(o, "score", 0.0)),
                        "bbox": getattr(o, "bbox", []),
                    }
                )

        return {
            "entities": entities,
            "relations": relations,
            "classifications": classifications,
            "objects": objects,
        }

File: integrations/sie_ts_langchain/src/extractors.ts
/**
 * SIE extraction tool for LangChain.js
 *
 * Provides extraction using SIE's extract endpoint:
 * - SIEExtractor: Extraction tool implementing LangChain Tool
 *
 * Returns entities, relations, classifications, and detected objects.
 *
 * @example
 * ```typescript
 * import { SIEExtractor } from "@superlinked/sie-langchain";
 *
 * const extractor = new SIEExtractor({
 *   baseUrl: "http://localhost:8080",
 *   model: "urchade/gliner_multi-v2.1",
 *   labels: ["person", "organization", "location"],
 * });
 *
 * const result = await extractor.invoke("John Smith works at Acme Corp in NYC");
 * const parsed = JSON.parse(result);
 * console.log(parsed.entities);
 * console.log(parsed.relations);
 * ```
 */

import { Tool } from "@langchain/core/tools";
import {
  type ExtractOptions,
  type ExtractResult,
  SIEClient,
  type SIEClientOptions,
} from "@superlinked/sie-sdk";

/**
 * Configuration options for SIEExtractor.
 */
export interface SIEExtractorParams {
  /**
   * URL of the SIE server.
   * @default "http://localhost:8080"
   */
  baseUrl?: string;

  /**
   * Extraction model name/ID.
   * @default "urchade/gliner_multi-v2.1"
   */
  model?: string;

  /**
   * Optional pre-configured SIEClient instance.
   * If provided, baseUrl and other connection options are ignored.
   */
  client?: SIEClient;

  /**
   * Labels to extract (entity types, relation types, or classification labels).
   * @default ["person", "organization", "location"]
   */
  labels?: string[];

  /**
   * Minimum confidence threshold (0-1).
   */
  threshold?: number;

  /**
   * Target GPU type for routing (e.g., "l4", "a100-80gb").
   */
  gpu?: string;

  /**
   * Request timeout in milliseconds.
   * @default 180000 (3 minutes)
   */
  timeout?: number;

  /**
   * Tool name for use in agents.
   * @default "sie_extract"
   */
  name?: string;

  /**
   * Tool description for use in agents.
   */
  description?: string;
}

/**
 * LangChain tool for extraction using SIE.
 *
 * Wraps SIEClient.extract() to implement the LangChain Tool interface
 * for use in agents and chains. Returns JSON with entities, relations,
 * classifications, and detected objects.
 *
 * @example
 * ```typescript
 * import { SIEExtractor } from "@superlinked/sie-langchain";
 *
 * // Direct usage
 * const extractor = new SIEExtractor({
 *   model: "urchade/gliner_multi-v2.1",
 *   labels: ["person", "organization", "location"],
 * });
 * const result = await extractor.invoke("John Smith works at Acme Corp");
 * const parsed = JSON.parse(result);
 *
 * // Use in an agent
 * import { ChatOpenAI } from "@langchain/openai";
 * import { createReactAgent } from "@langchain/langgraph/prebuilt";
 *
 * const agent = createReactAgent({
 *   llm: new ChatOpenAI(),
 *   tools: [extractor],
 * });
 * ```
 */
export class SIEExtractor extends Tool {
  name: string;
  description: string;

  private readonly model: string;
  private readonly labels: string[];
  private readonly threshold?: number;
  private _client: SIEClient | undefined;
  private readonly _ownsClient: boolean;
  private readonly baseUrl: string;
  private readonly clientOptions: SIEClientOptions;

  constructor(params: SIEExtractorParams = {}) {
    const toolName = params.name ?? "sie_extract";
    const toolDescription =
      params.description ??
      "Extract structured information from text. " +
        "Input should be text to analyze. " +
        "Returns JSON with entities, relations, classifications, and detected objects.";

    super({});

    this.name = toolName;
    this.description = toolDescription;

    const {
      baseUrl = "http://localhost:8080",
      model = "urchade/gliner_multi-v2.1",
      client,
      labels = ["person", "organization", "location"],
      threshold,
      gpu,
      timeout = 180_000,
    } = params;

    this.baseUrl = baseUrl;
    this.model = model;
    this.labels = labels;
    this.threshold = threshold;
    this._client = client;
    this._ownsClient = !client;

    this.clientOptions = {
      timeout,
      gpu,
    };
  }

  /**
   * Get or create the SIEClient.
   */
  private get client(): SIEClient {
    if (!this._client) {
      this._client = new SIEClient(this.baseUrl, this.clientOptions);
    }
    return this._client;
  }

  /**
   * Extract structured information from text.
   *
   * @param text - Text to extract from.
   * @returns JSON string with entities, relations, classifications, and objects.
   */
  async _call(text: string): Promise<string> {
    const extractOptions: ExtractOptions = {
      labels: this.labels,
    };
    if (this.threshold !== undefined) {
      extractOptions.threshold = this.threshold;
    }

    const result: ExtractResult = await this.client.extract(this.model, { text }, extractOptions);

    return JSON.stringify({
      entities: result.entities.map((e) => ({
        text: e.text,
        label: e.label,
        score: e.score,
        ...(e.start !== undefined && { start: e.start }),
        ...(e.end !== undefined && { end: e.end }),
      })),
      relations: result.relations.map((r) => ({
        head: r.head,
        tail: r.tail,
        relation: r.relation,
        score: r.score,
      })),
      classifications: result.classifications.map((c) => ({
        label: c.label,
        score: c.score,
      })),
      objects: result.objects.map((o) => ({
        label: o.label,
        score: o.score,
        bbox: o.bbox,
      })),
    });
  }

  /**
   * Close the underlying client connection.
   */
  async close(): Promise<void> {
    if (this._client && this._ownsClient) {
      await this._client.close();
    }
  }
}
